pkg/cloudprovider/providers/aws/instancecache.go

```diff
@@ -0,0 +1,308 @@
+		return nil, cloudprovider.InstanceNotFound
+	}
+
+	state := i.findLatestState(nil)
+	if state != nil {
+		return state, nil
+	}
+
+	return nil, cloudprovider.InstanceNotFound
+}
+
+func (i *cachedInstance) findLatestState(cachePolicy *CachePolicy) *ec2.Instance {
+	i.mutex.Lock()
+	defer i.mutex.Unlock()
+
+	if i.state != nil {
+		if cachePolicy == nil || cachePolicy.IsValid(i.stateTimestamp) {
+			return i.state
+		} else {
+			glog.V(2).Infof("triggering refresh because cache policy %s expired", cachePolicy.Name)
+		}
+	}
+
+	return nil
+}
+
+func newInstanceCache(cloud *Cloud) *instanceCache {
+	return &instanceCache{
+		cloud: cloud,
+	}
+}
+
+func (c *instanceCache) latestSnapshot() *instanceCacheSnapshot {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	return c.latest
+}
+
+func (c *instanceCache) refreshAll(previous *instanceCacheSnapshot) (*instanceCacheSnapshot, error) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	if c.latest != nil && c.latest != previous {
+		// A concurrent refresh happened, return that
+		return c.latest, nil
+	}
+
+	filters := []*ec2.Filter{
+		newEc2Filter("instance-state-name", "running"),
+	}
+
+	filters = c.cloud.addFilters(filters)
+	request := &ec2.DescribeInstancesInput{
+		Filters: filters,
+	}
+
+	instances, err := c.cloud.ec2.DescribeInstances(request)
+	if err != nil {
+		glog.V(2).Infof("error querying EC2 instances: %v", err)
+		return nil, err
+	}
+
+	if len(instances) == 0 {
+		glog.Warningf("did not find any running instances")
+	}
+
+	byInstanceID := make(map[awsInstanceID]*cachedInstance)
+	byNodeName := make(map[types.NodeName]*cachedInstance)
+
+	now := nanoTime()
+
+	latest := c.latest
+	if latest == nil {
+		// Easier than nil checks everywhere
+		latest = &instanceCacheSnapshot{}
+	}
+
+	for _, i := range instances {
+		instanceID := awsInstanceID(aws.StringValue(i.InstanceId))
+
+		if instanceID == "" {
+			glog.Warningf("ignoring instance without id: %v", i)
+			continue
+		}
+
+		if byInstanceID[instanceID] != nil {
+			return nil, fmt.Errorf("multiple instances found for instance: %s", instanceID)
+		}
+
+		ci := latest.byInstanceID[instanceID]
+		nodeName := mapInstanceToNodeName(i)
+
+		if ci == nil {
+			az := ""
+			if i.Placement != nil {
+				az = aws.StringValue(i.Placement.AvailabilityZone)
+			}
+
+			ci = &cachedInstance{
+				ID:               instanceID,
+				NodeName:         nodeName,
+				VpcID:            aws.StringValue(i.VpcId),
+				AvailabilityZone: az,
+				InstanceType:     aws.StringValue(i.InstanceType),
+				SubnetID:         aws.StringValue(i.SubnetId),
+
+				cache: c,
+			}
+		}
+
+		byInstanceID[instanceID] = ci
+		{
+			ci.mutex.Lock()
+			ci.state = i
+			ci.stateTimestamp = now
+			ci.mutex.Unlock()
+		}
+
+		if nodeName == "" {
+			glog.Warningf("ignoring instance with no NodeName: %s", instanceID)
+		} else {
+			if byNodeName[nodeName] != nil {
+				glog.Errorf("detected two instances with the same NodeName: %q", nodeName)
+			} else {
```

Comment:
Shouldn't we simply replace the instance even if another instance with same name was found?